src/hazelcore/dynamics/EvaluatorStep.re

```diff
@@ -0,0 +1,490 @@
+open Sexplib.Std;
+
+module EvalCtx = {
+  [@deriving sexp]
+  type t =
+    | Mark
+    | Ap1(t, DHExp.t)
+    | Ap2(DHExp.t, t)
+    | BinBoolOp1(DHExp.BinBoolOp.t, t, DHExp.t)
+    | BinBoolOp2(DHExp.BinBoolOp.t, DHExp.t, t)
+    | BinIntOp1(DHExp.BinIntOp.t, t, DHExp.t)
+    | BinIntOp2(DHExp.BinIntOp.t, DHExp.t, t)
+    | BinFloatOp1(DHExp.BinFloatOp.t, t, DHExp.t)
+    | BinFloatOp2(DHExp.BinFloatOp.t, DHExp.t, t)
+    | Cons1(t, DHExp.t)
+    | Cons2(DHExp.t, t)
+    | Pair1(t, DHExp.t)
+    | Pair2(DHExp.t, t)
+    | Let(DHPat.t, t, DHExp.t)
+    | Inj(HTyp.t, InjSide.t, t)
+    | NonEmptyHole(
+        ErrStatus.HoleReason.t,
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+      )
+    | Cast(t, HTyp.t, HTyp.t)
+    | FailedCast(t, HTyp.t, HTyp.t)
+    | InvalidOperation(t, InvalidOperationError.t)
+    | ConsistentCase(t, list(DHExp.rule), int)
+    | InconsistentBranches(
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+        list(DHExp.rule),
+        int,
+      );
+};
+
+let rec is_boxedval = (d: DHExp.t): bool =>
+  switch (d) {
+  | BoundVar(_)
+  | Lam(_, _, _)
+  | ListNil(_)
+  | BoolLit(_)
+  | IntLit(_)
+  | FloatLit(_)
+  | Triv => true
+  | Pair(d1, d2) => is_boxedval(d1) && is_boxedval(d2)
+  | Inj(_, _, d1) => is_boxedval(d1)
+  | Cons(d1, d2) => is_boxedval(d1) && is_boxedval(d2)
+  | Cast(d1, ty, ty') =>
+    switch (
+      EvaluatorCommon.ground_cases_of(ty),
+      EvaluatorCommon.ground_cases_of(ty'),
+    ) {
+    | (Ground, Hole)
+    | (Ground, NotGroundOrHole(_))
+    | (NotGroundOrHole(_), Ground) => is_boxedval(d1)
+    | (NotGroundOrHole(_), NotGroundOrHole(_)) =>
+      !HTyp.eq(ty, ty') && is_boxedval(d1)
+    | (Hole, Ground) =>
+      switch (d1) {
+      | Cast(_, _, Hole) => false
+      | _ => is_boxedval(d1)
```

Comment:
there aren't any values of type hole other than a cast, so we should raise an IllTyped exception in situations ruled out by that kind of reasoning